text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+efficiently produce more that this, e.g. the [ChaCha] algorithm can
+very naturally produce 128-bit `i32x4` SIMD vectors (and equally
+naturally produce up to 512-bit AVX512 vectors), and [dSFMT] naturally
+generates `f64x2` vectors. Future algorithms are likely to be designed
+for use with SIMD so it may be desirable to provide some way to
+special case more than just those 4 types.
+
+[ChaCha]: http://en.wikipedia.org/wiki/Salsa20#ChaCha_variant
+[dSFMT]: http://www.math.sci.hiroshima-u.ac.jp/~%20m-mat/MT/SFMT/index.html#dSFMT
+
+These methods are optional, and users are recommended to call
+`RngExt::gen` instead of these, so they are tentatively marked
+`unstable` in case an alternative approach is devised (they could also
+just be stabilised as is, since it is presumably desirable for the
+current four types to be efficient; any other types that we find that
+need this sort overloading can be added as default methods in future).
+
+
+#### Alternatives
+
+Inherit from `Iterator<u32>` instead of `Reader`, replacing `next_u32`
+with a substitute for `read` (the current `Rng` calls it
+`fill_bytes`). Most PRNG algorithms will find it annoying to implement
+`Reader` (they are defined in terms of `u32` or larger, not bytes),
+and, to retain efficiency, there may be some surprising behaviour with
+`Reader`, e.g.
+
+    let mut a = [0, 0];
+    rng.read(&mut a);
+
+    let mut b = [0, 0];
+    rng.read(&mut b[0..1]);
+    rng.read(&mut b[1..2]);
+
+will give different results. That said, this is inherently expected
+with randomness: a true random number generator can not be relied on
+at all.
+
+### The `RngExt` trait
+
+[the `RngExt` trait]: #the-rngext-trait
+
+`RngExt` provides basic useful functionality on RNG. Users will almost always
+want to `use` this trait (not `Rng`) into scope.
+
+`self.gen(data)` is a thin wrapper around `Random::random` as exists
+today, i.e. `Random::random(data, self)`.
+
+`rng.gen_iter(data)` returns an infinite iterator of random values,
+so that `for x in rng.gen_iter::<T>(data) { ... }` is like `loop { let
+x = rng.gen::<T>(data); ... }`. This method is unusual and consumes
+`self` so that the returned value is maximally flexible, e.g. it can
+be returned from the stack frame in which the RNG is created
+(currently it takes `&'a mut self` and returns a `Generator<'a, ...`),
+however non-consuming behaviour can be achieved by using the
+implementation of `Rng` for `&mut R` `R: Rng`.
+
+`rng.shuffle(values)` reorders the data in `values: &mut [T]`. If
+`rng` is a properly uniform RNG then any particular ordering occurs
+with probability `1/values.len()!`. This method could be more general:
+the restriction for the values to be contiguous in memory (`[T]`) is
+not necessary (e.g. it could perfectly well reorder the elements of a
+`RingBuf`), but `std` does not have the required abstraction to make
+it generic now. However, it is likely that it could be made generic in
+future, backwards-compatibly.
```

Comment:
I'd considered using `RandomAccessIterator`, but I'm not sure it is the right abstraction, e.g. I don't know how to use it to reorder values in a `&mut [T]`, and I'm not sure how one might implement it efficiently (specifically, not doing a length-n allocation) to give a perfectly even chance of any output order.

Also, unfortunately, it is `unstable` and so not usable in stable code.

One approach would be a trait capturing values which implement `swap(&mut self, i: usize, j: usize)`.